### PR TITLE
Encapsulate token request exchange between client and issuer

### PR DESCRIPTION
**Pull request by chris-wood**
Addresses #198. Starting as a draft for now since I don't have matching test vectors (yet).

### CODE REVIEW

File: draft-privacypass-rate-limit-tokens.md [hunk 1]
```diff
@@ -1261,9 +1346,9 @@ b2acb58dcfd23f54fc2084373cd9a78f7c84047c6f64e7464999597267e9b6a0f95bf13f
 44c4ec6abb1c846a16a222116ad94
 request_key: 8a9ffd91064e498f0b14b4b82d65255b2859da8d349e2bb362ca7c8211e
 23fcd5cd01813ef6eb4b994d532c3520e1aa89e
-origin_name_key_id:
+issuer_encap_key_id:
 590ef89ec14ebd7cd0598c7c1e045838e7c484713265c10d923d314fa1ef433f
-encrypted_origin_name: b9d946ee16d9114b6d42524a1fe1e5fae9768c45d4b48ce13
+encrypted_token_request: b9d946ee16d9114b6d42524a1fe1e5fae9768c45d4b48ce13
```
Comment (tlepoint) [inline draft-privacypass-rate-limit-tokens.md:1351]: This seems not to have been updated. Also we should delete the "blinded_msg" above, and add the issuer response?

Comment (chris-wood) [reply draft-privacypass-rate-limit-tokens.md:1351]: Yeah, I haven't updated any of the test vectors yet, but will do so before I mark this PR as ready. This particular change is the result of s/encrypted_origin_name/encrypted_token_request.

**Comment by chris-wood** [comment]
@tfpauly @tlepoint test vectors updated, based on [this change](https://github.com/cloudflare/pat-go/pull/11).

File: draft-privacypass-rate-limit-tokens.md [hunk 2]
```diff
@@ -268,9 +268,9 @@ Issuers MUST provide three parameters for configuration:
 1. Issuer Request URI: a token request URL for generating access tokens.
    For example, an Issuer URL might be https://issuer.example.net/token-request. This parameter
    uses resource media type "text/plain".
-1. Origin Name Key: a `NameKey` structure as defined below to use when encrypting the Origin
-   Name in issuance requests. This parameter uses resource media type "application/issuer-name-key".
-   The Npk parameter corresponding to the HpkeKdfId can be found in {{HPKE}}.
+1. Issuer Encapsulation Key: a `EncapsulationKey` structure as defined below to use when encapsulating
+   information to the Issuer in issuance requests. This parameter uses resource media type
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:272]:
```suggestion
   information, such as the origin name, to the Issuer in issuance requests. This parameter uses resource media type
```

File: draft-privacypass-rate-limit-tokens.md [hunk 3]
```diff
@@ -474,6 +474,15 @@ Its ABNF is:
     Sec-Token-Request-Blind = sf-binary
 ~~~
 
+The "Sec-Token-Request-Key" is an Item Structured Header {{!RFC8941}}. Its
+value MUST be a Byte Sequence. This header is sent on Client-to-Attester
+requests ({{request-one}}), and contains a per-request public key.
+Its ABNF is:
+
+~~~
+    Sec-Token-Request-Key = sf-binary
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:483]: I think we need to add this to IANA considerations?

Comment (chris-wood) [reply draft-privacypass-rate-limit-tokens.md:483]: Yep! Forgot to do that.

**Review by tfpauly — APPROVED** [review]

File: draft-privacypass-rate-limit-tokens.md [hunk 4]
```diff
@@ -519,8 +528,19 @@ blinded_msg, blind_inv = rsabssa_blind(pkI, token_input)
 The Client then uses Client Key to generate its one-time-use request public
 key `request_key` and blind `request_blind` as described in {{client-anon-issuer-origin-id}}.
 
-The Client then encrypts the origin name using Origin Name Key, producing
-`encrypted_origin_name` as described in {{encrypt-origin}}.
+The Client then constructs a OriginTokenRequest value, denoted `origin_token_request`,
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:531]:
```suggestion
The Client then constructs a InnerTokenRequest value, denoted `inner_token_request`,
```

File: draft-privacypass-rate-limit-tokens.md [hunk 5]
```diff
@@ -519,8 +528,19 @@ blinded_msg, blind_inv = rsabssa_blind(pkI, token_input)
...
+combining `blinded_msg`,  `request_key`, and the origin name as follows:
+
+~~~
+struct {
+  uint8_t blinded_msg[Nk];
+  uint8_t request_key[49];
+  uint8_t padded_origin_name<0..2^16-1>;
+} OriginTokenRequest;
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:539]:
```suggestion
} InnerTokenRequest;
```

File: draft-privacypass-rate-limit-tokens.md [hunk 6]
```diff
@@ -549,25 +567,22 @@ The structure fields are defined as follows:
...
-- "request_key" is computed as described in {{index-request}}.
+- "issuer_encap_key_id" is a collision-resistant hash that identifies the Issuer
+Encryption Key, generated as SHA256(EncapsulationKey).
 
-- "origin_name_key_id" is a collision-resistant hash that identifies the Origin Name Key,
-generated as SHA256(NameKey).
-
-- "encrypted_origin_name" is an encrypted structure that contains Origin Name,
-calculated as described in {{encrypt-origin}}.
+- "encrypted_token_request" is an encrypted structure that contains an OriginTokenRequest
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:573]:
```suggestion
- "encrypted_token_request" is an encrypted structure that contains an InnerTokenRequest
```

File: draft-privacypass-rate-limit-tokens.md [hunk 7]
```diff
@@ -644,14 +660,14 @@ compromise; see {{sec-considerations}} for additional about this channel.
...
-Token Keys and Origin Name Keys held by the Issuer.
-- The TokenRequest.encrypted_origin_name can be decrypted using the
-Issuer's private key (the private key associated with Origin Name Key), and matches
-an Origin Name that is served by the Issuer. This name might be the empty string "",
+- The TokenRequest.token_key_id and TokenRequest.issuer_encap_key_id correspond to known
+Token Keys and Issuer Encapsulation Keys held by the Issuer.
+- The TokenRequest.encrypted_token_request can be decrypted using the
+Issuer's private key (the private key associated with Issuer Encapsulation Key), and contains
+a valid OriginTokenRequest whose unpadded origin name matches an Origin Name that is served by
+the Issuer. The Origin name associated with the OriginTokenRequest value might be the empty string "",
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:668]:
```suggestion
a valid InnerTokenRequest whose unpadded origin name matches an Origin Name that is served by
the Issuer. The Origin name associated with the InnerTokenRequest value might be the empty string "",
```

File: draft-privacypass-rate-limit-tokens.md [hunk 8]
```diff
@@ -665,19 +681,22 @@ was invalid in the process.
 ## Issuer-to-Attester Response {#response-one}
 
 If the Issuer is willing to give a token to the Client, the Issuer decrypts
-TokenRequest.encrypted_origin_name to discover "origin". If this fails, the Issuer rejects
-the request with a 400 error. Otherwise, the Issuer validates and processes the token request
-with Issuer Origin Secret corresponding to the designated Origin as described in {{issuer-anon-issuer-origin-id}}.
-If this fails, the Issuer rejects the request with a 400 error. Otherwise, the output is
+TokenRequest.encrypted_token_request to discover a OriginTokenRequest value. If this fails,
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:684]:
```suggestion
TokenRequest.encrypted_token_request to discover a InnerTokenRequest value. If this fails,
```

File: draft-privacypass-rate-limit-tokens.md [hunk 9]
```diff
@@ -665,19 +681,22 @@ was invalid in the process.
...
+processes the token request with Issuer Origin Secret corresponding to the designated
+Origin as described in {{issuer-anon-issuer-origin-id}}. If this fails, the Issuer
+rejects the request with a 400 error. Otherwise, the output is
 index_key.
 
 The Issuer completes the issuance flow by computing a blinded response as follows:
 
 ~~~
-blind_sig = rsabssa_blind_sign(skP, TokenRequest.blinded_msg)
+blind_sig = rsabssa_blind_sign(skP, OriginTokenRequest.blinded_msg)
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:694]:
```suggestion
blind_sig = rsabssa_blind_sign(skP, InnerTokenRequest.blinded_msg)
```

File: draft-privacypass-rate-limit-tokens.md [hunk 10]
```diff
@@ -740,79 +760,146 @@ struct {
...
 - a selected combination of KDF, identified by kdfID, and AEAD, identified by aeadID.
 
 Beyond the key configuration inputs, Clients also require the following inputs defined
-in {{request-one}}: `token_key_id`, `blinded_msg`, `request_key`, and `origin_name_key_id`.
+in {{request-one}}: `token_key_id`, `blinded_msg`, `request_key`, `origin_name`, and
+`issuer_encap_key_id`.
 
-Together, these are used to encapsulate Origin Name (`origin_name`) and produce
-Encrypted Origin Name (`encrypted_origin`).
+Together, these are used to encapsulate an OriginTokenRequest and produce an encrypted token
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:788]:
```suggestion
Together, these are used to encapsulate an InnerTokenRequest and produce an encrypted token
```

File: draft-privacypass-rate-limit-tokens.md [hunk 11]
```diff
@@ -740,79 +760,146 @@ struct {
...
-1. Concatenate the values of aad, enc, and ct, yielding encrypted_origin_name.
+1. Concatenate the values of aad, enc, and ct, yielding encrypted_token_request.
 
 Note that enc is of fixed-length, so there is no ambiguity in parsing this structure.
 
 In pseudocode, this procedure is as follows:
 
 ~~~
-enc, context = SetupBaseS(pkI, "TokenRequest")
+enc, context = SetupBaseS(pkI, "OriginTokenRequest")
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:811]:
```suggestion
enc, context = SetupBaseS(pkI, "TokenRequest")
```

File: draft-privacypass-rate-limit-tokens.md [hunk 12]
```diff
@@ -740,79 +760,146 @@ struct {
...
 ~~~
 
-Issuers reverse this procedure to recover the (padded) Origin Name by computing the AAD as
-described above and decrypting encrypted_origin_name with their private key skI (the private
-key corresponding to pkI), and stripping off padding bytes. In pseudocode, this procedure
-is as follows:
+Issuers reverse this procedure to recover the OriginTokenRequest value by computing the AAD as
+described above and decrypting encrypted_token_request with their private key skI (the private
+key corresponding to pkI). The `origin_name` value is recovered from OriginTokenRequest.padded_origin_name
+by stripping off padding bytes. In pseudocode, this procedure is as follows:
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:830]:
```suggestion
Issuers reverse this procedure to recover the InnerTokenRequest value by computing the AAD as
described above and decrypting encrypted_token_request with their private key skI (the private
key corresponding to pkI). The `origin_name` value is recovered from InnerTokenRequest.padded_origin_name
by stripping off padding bytes. In pseudocode, this procedure is as follows:
```

File: draft-privacypass-rate-limit-tokens.md [hunk 13]
```diff
@@ -740,79 +760,146 @@ struct {
...
-             encode(49, request_key),
-             encode(32, origin_name_key_id))
-enc, context = SetupBaseR(enc, skI, "TokenRequest")
-origin_name, error = context.Open(aad, ct)
+             encode(32, issuer_encap_key_id))
+context = SetupBaseR(enc, skI, "TokenRequest")
+origin_token_request, error = context.Open(aad, ct)
+~~~
+
+The `OriginTokenRequest.blinded_msg` and `OriginTokenRequest.request_key` values, along
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:845]:
```suggestion
The `InnerTokenRequest.blinded_msg` and `InnerTokenRequest.request_key` values, along
```

File: draft-privacypass-rate-limit-tokens.md [hunk 14]
```diff
@@ -740,79 +760,146 @@ struct {
...
+with the unpadded `origin_name` value, are used by the Issuer as described in {{request-two}}.
+
+## Issuer to Client Encapsulation {#encap-issuer-to-client}
+
+Given an HPKE context `context` computed in {{encap-client-to-issuer}}, Issuers encapsulate
+their token response `blind_sig`, yielding an encrypted token response `encrypted_token_response`,
+to the Client as follows:
+
+1. Export a secret `secret` from `context`, using the string "OriginTokenResponse" as context.
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:854]:
```suggestion
1. Export a secret `secret` from `context`, using the string "TokenResponse" as context.
```

File: draft-privacypass-rate-limit-tokens.md [hunk 15]
```diff
@@ -740,79 +760,146 @@ struct {
...
+   `nonce`, empty `aad`, and a `pt` input of `request`, which yields `ct`.
+
+7. Concatenate `response_nonce` and `ct`, yielding an Encapsulated Response
+   `enc_response`. Note that `response_nonce` is of fixed-length, so there is no
+   ambiguity in parsing either `response_nonce` or `ct`.
+
+In pseudocode, this procedure is as follows:
+
 ~~~
+secret = context.Export("OriginTokenResponse", Nk)
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:883]:
```suggestion
secret = context.Export("TokenResponse", Nk)
```

File: draft-privacypass-rate-limit-tokens.md [hunk 16]
```diff
@@ -1222,49 +1309,52 @@ The test vector below for the procedure in {{encrypt-origin}} lists the followin
...
-- origin_name_key: The public Origin Name Key, represented as a hexadecimal string.
+- issuer_encap_key: The public Issuer Encapsulation Key, represented as a hexadecimal string.
 - token_type: The type of the protocol specified in this document, 0x0003.
 - token_key_id: The ID of Token Key computed as in {{request-one}}, a single octet.
 - blinded_msg: A random blinded_msg value, represented as a hexadecimal string.
 - request_key: A random request_key value, represented as a hexadecimal string.
-- origin_name_key_id: The Origin Name Key ID computed as in {{request-one}}, represented as a hexadecimal string.
-- encrypted_origin_name: The encrypted Origin Name, represented as a hexadecimal string.
+- issuer_encap_key_id: The Issuer Encapsulation Key ID computed as in {{request-one}}, represented as a hexadecimal string.
+- encrypted_token_request: The encrypted OriginTokenRequest, represented as a hexadecimal string.
```
Comment (tfpauly) [inline draft-privacypass-rate-limit-tokens.md:1321]:
```suggestion
- encrypted_token_request: The encrypted InnerTokenRequest, represented as a hexadecimal string.
```

**Review by dvorak42 — APPROVED** [review]

**Pull request merged**